[IV-23-24] Objetivo 2

puchy22 opened the pull request:
**Lista de comprobación:**
*Sobre la estructura del repositorio:*
* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas específicas para el lenguaje con el que se está trabajando?
*Sobre el planteamiento:*
* [x] ¿Se han planteado una serie de issues, asignados a las historias de usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?
*Sobre el análisis del problema:*
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor, una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo objeto valor que no dependa de ningún otro (y que sea la base de muchos otros)?.
*Sobre el código*
* [x] ¿Todos los mensajes de commit explican el cambio, y no se limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [ ] ¿Ha asignado el propietario/la propietaria del repositorio todos los issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto mínimamente viable? ¿O tengo que solicitar al product manager que precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente correcto?

@shvtwp he añadido la lista de comprobación rellena que antes se me había olvidado añadirla en los mensajes, creo que tu tendrías que realizar la otra para considerar que los cambios realizados son los oportunos.

puchy22 committed:
Elegido el lenguaje de programación por consenso
añadido al fichero iv.yaml y gitignore correspondiente al lenguaje typescript closes #8

puchy22 committed:
Añadida la entidad horario

puchy22 committed:
Añadida la clase encargada de cuidar a una persona
en esta se implementa lo que se define en las HUs como un familiar que cuida a una persona close #14 close #11

puchy22 committed:
Añadida la entidad que modela al Horario
en este fichero se implementa la clase que almacenará a la persona a atender y una lista con todos sus cuidadores disponibles close #13

puchy22 commented:
@shvtwp Listo para revisión. Si ves que las estructuras definidas están bien apruebala y ya aviso a JJ para que lo revise.

puchy22 commented:
**Lista de comprobación:**
*Sobre la estructura del repositorio:*
* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas específicas para el lenguaje con el que se está trabajando?
*Sobre el planteamiento:*
* [x] ¿Se han planteado una serie de issues, asignados a las historias de usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?
*Sobre el análisis del problema:*
* [ ] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo que se ha creado?
* [ ] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor, una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo objeto valor que no dependa de ningún otro (y que sea la base de muchos otros)?.
*Sobre el código*
* [x] ¿Todos los mensajes de commit explican el cambio, y no se limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [ ] ¿Ha asignado el propietario/la propietaria del repositorio todos los issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto mínimamente viable? ¿O tengo que solicitar al product manager que precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente correcto?

shvtwp commented:
* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [ ] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
@puchy22 Aquí tengo una última duda, ¿no debería tener un horario semanal una representación de los días de la semana? Quizá lo hayas pensado de otra forma o consideres que es más tarea mía, pero me gustaría saber qué piensas.

* [ ] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

puchy22 commented:
@shvtwp la forma de abordar el problema que yo consideré es que al final la salida de tu lógica de negocio es una representación de una serie de horas y las asignaciones a unos cuidadores pero esa estructura que almacena las horas es parte de la solución por lo que yo consideré que se podía generar al inicio del algoritmo en función de lo longevo que quieras hacer el horario (una semana o un año).

Aunque la solución que tú das también es válida, se puede inicializar un array en el constructor de la clase Horario donde ir almacenando las fechas de inicio y fin y el id de la persona que cuidará durante ese tramo horario. Al final la única diferencia está en donde se genera esa estructura de datos.

Yo opté por la solución donde el array con los tramos horarios y las asignaciones se genera conforme avanza la lógica de negocio por pensar que era más minimalista, pero es verdad que de primeras ver una clase Horario que no tiene ninguna estructura de datos para almacenar Fechas es un poco contraintuitivo.

Comentadas estas dos visiones que tenía sobre el problema comentame como lo prefieres y en función de eso añado la estructura de datos que te he comentado, si tienes otra manera de plantearlo que no había tenido en cuenta o si lo apruebas así y le pido revisión a JJ.

Un saludo :wave:

shvtwp commented:
@puchy22 Entendido, me parece bien. ¿Debes dejar el PR como borrador?

puchy22 commented:
@shvtwp ya está lista la PR para ser revisada, gracias por recordarmelo :+1:

puchy22 commented:
@JJ Listo para revisión.

shvtwp reviewed (approved):
* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [ ] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
* [ ] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

JJ reviewed (changes requested):
No se considera buena práctica usar ficheros con mayúsculas en ningún repo. Ya lo comenté en clase el viernes pasado y la razón.

puchy22 committed:
Cambios en el nombre de los ficheros de código
se han modificado el nombre de los ficheros con el código typescript para que no incluyan mayúsculas

puchy22 commented:
@JJ ya he cambiado el nombre de los ficheros para que no incluyan mayúsculas.

JJ commented on src/cuidador.ts:
```diff
@@ -0,0 +1,25 @@
+   NO_TRABAJADOR
+}
+
+export class Cuidador{
```
El commit que ha creado esto https://github.com/shvtwp/DePendiente/commit/84dfada43632031042b4de60b5dc4079464b68ab no sigue las mejores prácticas, pero además referencia un issue que no está relacionado con ninguna HU y ese issue no hace ninguna referencia al Cuidador, sólo al intervalo horario.

JJ reviewed (changes requested):
Por favor, tratad de seguir las buenas prácticas tanto en la creación de issues como en los mensajes de commit.
En este caso, la forma correcta de proceder es reescribir el código, aparte por supuesto de editar el issue. Si el código se ha escrito por un issue que no está relacionado con el mismo, sólo se puede borrar porque no aporta valor al cliente.
Por otro lado, consultad los [errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-código-definido), sobre todo en el apartado que habla de cómo diferenciar las partes mutables de las partes inmutables de un objeto en un modelo.

puchy22 committed:
Eliminada la interfaz IntervaloHorario
se ha eliminado porque su issue no tenía HUs asociadas close #14

puchy22 committed:
Definición del objeto valor que define Cuidador
se define el objeto valor Cuidador planteado en las distintas HUs close #11

puchy22 committed:
Definida las preferencia para la asignación de horas
se solventa el problema de preferencia de asignación de horas definido en la #6

puchy22 commented:
@JJ. Listo para revisión.

JJ commented on src/horario.ts:
```diff
@@ -0,0 +1,20 @@
+
+class Horario{
+
+   readonly PORCENTAJE_DE_PREFERENCIA_ESTUDIANTE = 0.5;
```
Esto deberían ser constantes; puedes definirlas fuera de la clase.

JJ commented on src/horario.ts:
```diff
@@ -0,0 +1,20 @@
+
+   readonly PORCENTAJE_DE_PREFERENCIA_ESTUDIANTE = 0.5;
+   readonly PORCENTAJE_DE_PREFERENCIA_TRABAJADOR = 0.75;
+   readonly PORCENTAJE_DE_PREFERENCIA_NO_TRABAJADOR = 0.85;
```
Tienes que crear un issue; no puedes referenciar directamente la historia de usuario.

puchy22 committed:
Cambio en las condiciones por el issue #15
se ha cambiado la condición estudiante por joven, por ser más fiel con la HU

JJ reviewed (changes requested):
Dale una vueltecilla, por favor.

puchy22 committed:
Cambio representación equidad en OV persona cuidadora(#15)
en el caso de las personas cuidadoras he interpretado que el único dato
relevante es si se trata de una persona joven como se indica explícitamente
en la HU2, por lo que se han eliminado demás condiciones al ser
calculables con el horario disponible de cada persona

puchy22 committed:
Representación equidad en entidad Horario close #15
creado un método a implementar en el que cuando sea necesario calcular en
un tramo del horario que persona es la más prioritaria devuelva como resultado
una persona que debe ser seleccionada de acuerdo con el número de horas libres,
definido en otro método a implementar y si es Joven o no, característica almacenada
en cada persona cuidadora

puchy22 commented:
@JJ Listo para revisión. Cambios realizados a partir de lo que hemos hablado en clase sobre la representación de la equidad.

JJ commented:
Creo que has hecho suficientes cambios como para necesitar otra revisión de @shvtwp  antes de que yo haga la mía.

puchy22 commented:
@shvtwp Listo para revisión. Los cambios más significativos han sido alrededor de como se modelaría la equidad indicada en la HU2. Si estas conforme dímelo por aquí por un comentario, en caso contrario solicita los cambios que veas pertinentes.

puchy22 commented:
@JJ Listo para revisión.

JJ commented on persona_cuidadora.ts:
```diff
@@ -0,0 +1,19 @@
+
+interface IntervaloHorario {
```
¿Seguro que este fichero está en el lugar correcto?

JJ commented on src/horario.ts:
```diff
@@ -0,0 +1,25 @@
+   private idPersonaAtendida: number;
+   private personasCuidadorasDisponibles: Array<PersonaCuidadora>;
+
+   constructor(idPersonaAtendida: number){
```
Ya enlacé desde el grupo de Telegram el método abreviado para declarar constructores. Mira, por otro lado, [los errores frecuentes en el código definido](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-c%C3%B3digo-definido) sobre la necesidad de incluir o no dicho constructor (si no se usa el método abreviado de TS) o los getters.

JJ reviewed (changes requested):
Adicionalmente a lo comentado, por favor mirad cuál es la buena práctica sobre el nombre del directorio en el que se almacenan los ficheros.

puchy22 committed:
Cambio ubicación del fichero de personas cuidadoras
se ha transladado el fichero de la clas de personas cuidadoras a la carpeta de los ficheros fuentes

puchy22 committed:
Eliminado el constructor de la entidad horario
se ha eliminado el constructor de la clase horario por no aportar valor a la
modelización del problema

puchy22 committed:
Cambio en el nombre del directorio src
se ha cambiado el directorio src a lib para seguir unas mejores prácticas en el
proyecto, ya que lib en proyectos TypeScript suele contener piezas independientes
de código

puchy22 commented:
@JJ listo para revisión. Se han corregido y documentado en los commits los cambios recomendados.

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: TypeScript
+entidad: src/horario.ts
```
Ya no está en ese path.

JJ commented on lib/horario.ts:
```diff
@@ -0,0 +1,20 @@
+class Horario{
+
+   private idPersonaAtendida: number;
+   private personasCuidadorasDisponibles: Array<PersonaCuidadora>;
```
¿Has mirado la forma abreviada que comenté en clase de declarar las variables _y_ el constructor en TS?

puchy22 replied:
Sí que lo he visto pero como en los errores frecuentes se comenta que si el constructor no aporta nada y solo se asignan los parámetros del mismo no se pongan, he entendido que no era necesario añadir constructor en este caso.

JJ replied:
Con la forma abreviada declaras el constructor *y* las variables.

JJ commented on lib/horario.ts:
```diff
@@ -0,0 +1,20 @@
+   private idPersonaAtendida: number;
+   private personasCuidadorasDisponibles: Array<PersonaCuidadora>;
+
+   private getNumeroHorasLibres(personaCuidadora: PersonaCuidadora): number{
```
En los [errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-c%C3%B3digo-definido) comenta sobre los getters en el mismo lugar donde comenta sobre el constructor. En este caso, la elección que has hecho es que esto va a ser parte del interfaz de *esta* clase y no una consecuencia del horario. Si no tienes claro cómo se va a calcular esto, ni a partir de qué (si de los datos contenidos en esta clase u otra) es mejor que no lo incluyas.

JJ commented on lib/horario.ts:
```diff
@@ -0,0 +1,20 @@
+      return 0;
+   }
+
+   public elegirPersonaCuidadora(personasCuidadoras: Array<PersonaCuidadora>): PersonaCuidadora{
```
No entiendo muy bien cómo estas dos funciones representan la equidad. Para empezar, no lo representan, porque simplemente no aparece en ninguno de los nombres. Para seguir, las estás incluyendo en una clase `Horario` que es el horario *para una sola persona*. Si la persona cuidadora cuida a varias personas, ¿no debería una clase tenerlo en cuenta? Si todo esto se está haciendo para cuidar a una sola persona, ¿para qué se necesita el ID?
Si se va a usar el concepto de equidad, debería de exponerse todo el contexto en #15, cosas como quizá la historia de asignaciones (¿se pide en #6?)

puchy22 replied:
En mi planteamiento todo se hace solo para cuidar a una sola persona ya que en ningún momento se menciona en las HUs que haya que gestionar varios familiares. Si este fuese el caso no creo que hubiera problema ya que para asignarse un segundo horario el usuario tendría que indicar que esas horas donde está cuidando al segundo familiar no son horas libres y para ese segundo horario tendría un ID que no tendría porqué ser igual al del otro horario, es ID iría ligado a ese Horario, no está pensado como un ID universal para todos los horarios.

Para explicarme mejor yo entiendo que la lógica de negocio como salida dará una asignación de un Tramo Horario a una Persona, por lo tanto será necesario identificarla de alguna manera como puede ser su DNI, ya que el nombre o apellidos podrían repetirse y al considerar a las Personas OV si el indetificador coincide significa que son lo mismo cuando deberían no serlo. Además justo en el caso de cuidar familiares es más probable que un nombre con unos apellidos se repitan.

Y en cuanto a como representa la equidad creo que el nombre se podría mejorar, pero lo que yo había entendido por equidad es que en algún punto del algoritmo para cada tramo horario iremos seleccionando personas usando como criterio las horas libres, por lo que el cómputo total de horas libres que tendrá esa persona restantes hasta la fecha máxima que se calcule el horario final irá variando. Por lo que ese método representa la equidad en el sentido de que elegirá a la persona que más horas libres tenga en ese punto de la asignación del algoritmo, por lo que pienso que la estoy planteando el problema de manera que se realizase una solución del tipo Greedy y que esta función devolvería los óptimos locales. No sé si sería la solución correcta pero es como yo había pensado en un inicio que se usaría este método y como se resolvería el tema de la equidad.

Es cierto que esto no había tenido en cuenta la representación de que un familiar debería cumplir un máximo de horas por día, que ya lo he añadido en los recientes commits.

JJ reviewed (changes requested):
Para ver si realmente va a ser útil en el objetivo 4, aconsejo tanto a @puchy22 como a @shvtwp que hagan un pequeño "recorrido" por el código existente para ver cómo se llevaría a cabo la implementación de la lógica de negocio:

"Se recorren los horarios disponibles de cada  uno. A continuación se toma... " 

Para ver si efectivamente es útil, tendrían que estar prácticamente todos los elementos necesarios programados. En este caso, aparte del hecho de que los horarios disponibles estén, no parece que haya mucho más donde agarrarse. Por ejemplo, no parece muy claro que la implementación de #13 contemple que no se pasan demasiadas horas seguidas, ni la necesidad de un profesional. ¿En qué parte del modelo se está contemplando eso?

puchy22 committed:
Cambiada la ruta de la entidad

puchy22 committed:
Limitación de horas para personas jóvenes
se limita la cantidad de horas diarias que una persona joven puede cuidar a 2 horas

puchy22 committed:
Añadido límite de horas por descanso #13
se añade un límite máximo de horas diaris que un cuidador puede pasar

puchy22 committed:
Cambio en los constructores
se han añadido constructores a ambas clases con el uso de una sintaxis más breve
y más correcta para TypeScript

puchy22 committed:
Añadida representación de horas de un profesional #13
se ha añadido la declaración de un método el cual asigna un tramo horario a un profesional en caso
de no exisitir ningún familiar disponible para ese tramo horario

puchy22 committed:
OV persona cuidadora a entidad familiar cuidador close #11
cambio en el concepto de de una persona cuidadora pasa a ser una entidad debido a que
durante la lógica de negocio las horas disponibles pueden ir variando en función de
la asignación que se vaya haciendo, por lo que considero que son una entidad que necesita
poder ser almacenada e identificada más allá de sus valores

puchy22 committed:
Añadido OV turno de cuidado close #14
se añade el OV valor que representa los distintos turnos de cuidado que puede tener
libre un familiar cuidador, y también son usados para componer el horario final
fruto de aplicar la lógica de negocio

puchy22 committed:
Cambio representación entidad Horario #13
se han cambiado los valores que definen a una entidad horario para que se ajusten
más a la descripción proporcionada por las historias de usuario y sean más útiles
a la hora de trabajar en la implementación de la lógica de negocio, por lo que se
añade un map el cual asigna el id de un familiar a un determinado turno de cuidado
además se añade la representación de la asignación a profesionales como una serie
de turnos de cuidado

puchy22 committed:
Equidad en horario #13 close #15
se ha añadido el método que durante la lógica de negocio debe elegir a que persona
toca asignar su tramo horario libre, este se hace de manera equitativa ya que
escoge siempre a la persona que la suma de las horas de sus tramos de cuidado
disponibles sea mayor

puchy22 committed:
Limitacion familiares jovenes close #13
añadida limitación de horas máximas para los familiares jóvenes

puchy22 commented:
@shvtwp revisa los cambios que he hecho ya que creo que son significativos, además he actualizado la lista de comprobación y añadido al cuerpo del PR. Si lo ves bien puedes aprobarlo y solicitarle la revisión a @JJ. Si tienes alguna sugerencia a cambiar indicamela sin problema, los cambios hechos se documentan todos en los últimos commits. Un saludo :wave:

shvtwp reviewed (approved):
Considero que ahora la representación es mucho más clara, creo que está todo correcto.

puchy22 commented:
@JJ Listo para revisión.

JJ reviewed (approved):
:+1: gracias por los cambios hechos; por favor en el futuro trata de hacer que los commits sigan las buenas prácticas sin repetir lo que ya está claro por la propia estructura del cambio.
